torchrl/envs/libs/openspiel.py

```diff
@@ -0,0 +1,358 @@
+
+    @_classproperty
+    def available_envs(cls):
+        if not _has_pyspiel:
+            return []
+        return _get_envs()
+
+    def __init__(self, env=None, categorical_action_encoding=True, **kwargs):
+        if env is not None:
+            kwargs["env"] = env
+        self._seed_calls_reset = None
+        if not categorical_action_encoding:
+            raise NotImplementedError
+        self._categorical_action_encoding = categorical_action_encoding
+        super().__init__(**kwargs)
+
+    def _check_kwargs(self, kwargs: Dict):
+        pyspiel = self.lib
+        if "env" not in kwargs:
+            raise TypeError("Could not find environment key 'env' in kwargs.")
+        env = kwargs["env"]
+        if not isinstance(env, pyspiel.State):
+            raise TypeError("env is not of type 'pyspiel.State'.")
+
+    def _build_env(self, env, requires_grad: bool = False, **kwargs):
+        self.requires_grad = requires_grad
+        return env
+
+    def _init_env(self):
+        self._update_action_mask()
+
+    def _make_specs(self, env: "pyspiel.State") -> None:  # noqa: F821
+        self.reward_spec = UnboundedContinuousTensorSpec(
+            shape=[*self.batch_size, env.num_players()],
+            device=self.device,
+        )
+        self.done_spec = DiscreteTensorSpec(
+            n=2,
+            shape=torch.Size((1,)),
+            dtype=torch.bool,
+            device=self.device,
+        )
+        action_spec_cls = (
+            DiscreteTensorSpec
+            if self._categorical_action_encoding
+            else OneHotDiscreteTensorSpec
+        )
+        self.action_spec = action_spec_cls(
+            env.num_distinct_actions(),
+            dtype=torch.int64,
+            device=self.device,
+        )
+        self.observation_spec = CompositeSpec(
+            observation=UnboundedContinuousTensorSpec(
+                shape=(
+                    *self.batch_size,
+                    env.num_players(),
+                    *env.get_game().observation_tensor_shape(),
+                ),
+                device=self.device,
+            ),
+            current_player=DiscreteTensorSpec(
+                env.num_players(),
+                dtype=torch.int,
+                device=self.device,
+            ),
+            state=NonTensorSpec([]),
+            shape=self.batch_size,
+        )
+
+    def _set_seed(self, seed):
+        if seed is not None:
+            raise NotImplementedError("This environment has no seed.")
+
+    def current_player(self):
+        return self._env.current_player()
+
+    def _update_action_mask(self):
+        action_mask = torch.zeros(
+            self._env.num_distinct_actions(), device=self.device, dtype=torch.bool
+        )
+        action_mask[self._env.legal_actions()] = True
+        self.action_spec.update_mask(action_mask)
+
+    def _make_td_out(self, exclude_reward=False):
+        observation_shape = self._env.get_game().observation_tensor_shape()
```

Comment:
How expensive is `get_game()`? Can we cache it?

Reply:
It isn't very expensive, but I tried caching it with `functools.lru_cache`, and it does speed it up by about 6x, so I'll update it.

```python
import timeit
iters = 1_000_000
res0 = timeit.timeit(
    setup='import torchrl; env = torchrl.envs.OpenSpielEnv("chess")',
    stmt='env._env.get_game()',
    number=iters)
res1 = timeit.timeit(
    setup='import torchrl; env = torchrl.envs.OpenSpielEnv("chess")',
    stmt='env._get_game_cached()',
    number=iters)
print(f'nocache: {res0 / iters} s')
print(f'cache: {res1 / iters} s')
print(f'speedup: {res0 / res1} x')
```

```
nocache: 4.2987480599913395e-07 s
cache: 6.127979099983349e-08 s
speedup: 7.014952221366127x
```

Reply:
Well, flake8 fails, pointing out that `lru_cache` on an instance method can cause memory leaks: https://docs.astral.sh/ruff/rules/cached-instance-method/

So I decided to just cache it manually using an initially-None data member. Speedup decreased to 5x, but that still makes it worth doing.